### ISSUE TITLE
Safari share links aren't copied to clipboard

### ISSUE THREAD

**Issue by maxwofford** [hackclub/sprig#15]

**Comment by maxwofford** [comment]
<img width="1433" alt="Screen Shot 2022-01-26 at 10 13 50" src="https://user-images.githubusercontent.com/5891442/151190398-4b9a0cf6-7bae-4287-a6ff-bd520aa4c5b1.png">

The share link (1) button should create a share link and add it to the user's clipboard. On safari it notifies the user that the link was added to their clipboard (2), but when pasting the share link doesn't come out (whatever was previously in the clipboard does).

**Comment by maxwofford** [comment]
Now running into this on FF
<img width="1139" alt="Screen Shot 2022-01-26 at 18 37 43" src="https://user-images.githubusercontent.com/5891442/151265057-4b804491-cbc1-4b01-a069-d57a7061e832.png">

**Comment by cjdenio** [comment]
Yeah, most browsers only let you copy to the clipboard as a direct result of user interaction.

Chrome lets you explicitly request permission via the [Permissions API](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API), but Firefox and Safari don't support it yet

**Comment by zachlatta** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand

**Comment by zachlatta** [comment]
So I've done a little research here, and it turns out the issue is that Safari has some security policies in place that prevent any clipboard commenting to happen unless a DOM event is in scope (so to ensure that a DOM event triggered the copy to clipboard).

I am not sure what the appropriate way of passing the DOM event to the copy function would be with the architecture of Game Lab. @leomcelroy / @cedric-h: this might be a good thing for one of you to handle.

I've tried using various 3rd party clipboard libraries and none of the ones I've tried work, so I'm pretty sure it's down to the DOM event thing.

It looks like this might be a very recent thing. There are some other repos with this issue popping up, notably https://github.com/apache/superset/issues/18154.

**Comment by maxwofford** [comment]
Copying doesn't work in any browser for me. Reopening

**Comment by maxwofford** [comment]
The diff for https://github.com/hackclub/gamelab/pull/67/files doesn't make sense to me– seems like something was mis-commited or something. If I'm reading it right it seems like it [removes copy logic](https://github.com/hackclub/gamelab/pull/67/files#diff-579aaf136d46b663d22c6d5e8aca317c5d13ad3c109ff721b7c18d90c70cfb7dL25-R25).

**Comment by faisalsayed10** [comment]
@maxwofford @leomcelroy this is my main commit - [`a7618a3` (#67)](https://github.com/hackclub/gamelab/pull/67/commits/a7618a3a800ca2ed8a4311809a0b0644c3d9f2d4)
i think something has went wrong, and we've mis-commited something

**Comment by faisalsayed10** [comment]
also @leomcelroy this commit of yours kinda reverted my commit [`b50ee51` (#67)](https://github.com/hackclub/gamelab/pull/67/commits/b50ee516b920199f9432cf6bf2b68a9f78fca7bb)

**Comment by maxwofford** [comment]
Something that might help for fixing this issue: saves are deterministic. Even before we make a post request, we know the share link will be gamelab.hackclub.com/?id=MD5 where MD5 is the [hash of the game state](https://github.com/hackclub/gamelab/blob/main/save.js#L5)

CC'ing @cedric-h who can correct me if I said something wrong.

**Comment by isabellalacerda** [comment]
Copying also doesn't work in any browser for me. How can i fix this?

![image](https://user-images.githubusercontent.com/102618441/176012010-7a88eb78-73b7-4c2e-8862-37f6368031b7.png)
![image](https://user-images.githubusercontent.com/102618441/176012273-ca696f39-95aa-4d2c-99d9-02dd364752b4.png)

**Comment by maxwofford** [comment]
Can you share the URL that screenshot was taken at? It doesn't look like a part of the gamelab interface

**Comment by Peiprjs** [comment]
You can copy the URL from the console. Otherwise, it only copies some Javascript

**Comment by alhardwarehyde** [comment]
@grymmy can you reproduce this?

**Comment by alhardwarehyde** [comment]
Ignore my ask. This is an ongoing issue

**Comment by grymmy** [comment]
An update on this issue - while the editor still does not correctly copy to the clipboard on safari, it shows a high-quality error and feedback to the user when it files to copy to the clipboard.  As a result, I believe this is quite low priority - but I've confirmed yes, still a present bug
